### DISCUSSION TITLE
感觉可以给字段模型新增一个data字段，专门用于扩展状态

### DISCUSSION THREAD

**Discussion by janryWang** [alibaba/formily#1590] [RFC]
大家怎么看？

**Comment by liuweiGL** [comment]
我这边没有碰到这种需求，希望可以说一下使用场景啊。

**Reply by janryWang** [reply]
比如下拉框搜索，需要维护一个局部keyword状态，这个是属于业务层的私有状态，用data来扩展感觉比较合适

**Reply by liuweiGL** [reply]
我现在的业务逻辑按照你之前的建议全部都放在组件里面来做了，这种状态在业务组件里面来维护就好了吧，还需要放到 `field` 中去？感觉不需要吧。

**Comment by JustDs** [comment]
同意，这一能力在业务自定义组件层的情况下是很有用的。

尤其是在数组场景，如果组件找不到合适的字段做 key 而只能用 index 做 key 的话，调整排序时会导致组件上的状态滞留。如果能把状态挂在 Field 上可以很有效地解决这一问题。

我甚至觉得 Form 上也应该有这样做状态扩展的能力，在对表单整体做封装时有时会需要扩展自定义状态。

**Reply by liuweiGL** [reply]
给个 demo 看看呀。

**Comment by yuyi919** [comment]
正好在翻文档没找到这个
1.x的时候我自定义了一个frozen字段的操作
更新状态并将更新前的原状态存在自定义的属性中，等到frozen解除后字段会恢复回原来的状态

**Reply by janryWang** [reply]
我在想这个扩展状态，要不要支持属性传入初始值，如果要支持的话，协议层可能又得加一个x-data来承载

**Comment by MatriXiao88** [comment]
@liuweiGL 
其实这个需求场景还是有的，在schemaForm场景下，因为有些表单字段上下文相关的数据，例如对于一个用户选择框，但业务上对于用户的额外描述信息，例如身高、血型，有时候比较定制的场景就是后端返回一个字段来表示是否需要提交这些额外描述信息，像这个开关变量就属于上下文数据，但目前没有找到怎么在创建schema时缓存该开关变量，然后在提交阶段再读取该开关变量去判断

如果有了扩展的data字段，对于这种字段上下文数据的缓存就有了去处了

**Reply by MatriXiao88** [reply]
现在对于component、decorator属性，都有对应的x-component-props、x-decorator-props去设置和缓存额外属性数据，但对于用于描述Field本身的额外上下文数据（业务上除了用于生成动态表单字段数据外，还会附带不少字段相关的其他数据），目前就没有办法在整个表单生命周期去传递

期待作者考虑加入这个feature

**Reply by UlricGan** [reply]
同感，有时候需要根据表单字段以外的数据进行联动（或是说联动依赖这些数据），比如字段B的默认值来源于字段A（假设是 select）的option 中的某个字段
PS：options 没有在 dataSource 里

**Comment by s330526719** [comment]
这个特性会支持吗？举个场景case：select组件的数据源设置为业务枚举字典。
预期字段schema定义：
```
"select": {
  "type": "string",
  "title": "下拉框-单选",
  "x-decorator": "FormItem",
  "x-component": "Select",
  "x-component-props": {
    "placeholder": "请选择",
    "allowClear": true
  },
  "x-data": {
    "enum-key": "enum_key1"
  },
  "x-validator": [{ "required": true, "message": "请选择下拉框-单选" }]
}
```
业务代码根据enum-key找到实际数据源，设置给field.dataSource：
```
onFieldInit('*', field => {
  // 已提前获取表单所需的全部枚举字典allEnums
  // const allEnums = {
  //   'enum_key1': [
  //     {label: '1', value: 1}
  //   ]
  // };
  const enumKey = field?.['x-data']?.['enum-key'];
  if (allEnums?.[enumKey]) {
    field.dataSource = allEnums?.[enumKey];
  }
}
```
这里enum-key并不需要被组件消费，同时可以给select、radio、checkbox等全部需要数据源的组件使用；可视化配置时，从全局枚举字典中选择后得到enum-key: '选择的枚举字典标识'。
